plugins/UM3NetworkPrinting/src/Network/NetworkOutputDeviceManager.py

```diff
@@ -0,0 +1,205 @@
+            b"manual": b"true",
+            b"incomplete": b"true",
+            b"temporary": b"true"
+        })
+        CuraApplication.getInstance().callLater(callback, True, address)
+
+    ## Returns a dict of printer BOM numbers to machine types.
+    #  These numbers are available in the machine definition already so we just search for them here.
+    @staticmethod
+    def _getPrinterTypeIdentifiers() -> Dict[str, str]:
+        container_registry = CuraApplication.getInstance().getContainerRegistry()
+        ultimaker_machines = container_registry.findContainersMetadata(type="machine", manufacturer="Ultimaker B.V.")
+        found_machine_type_identifiers = {}  # type: Dict[str, str]
+        for machine in ultimaker_machines:
+            machine_bom_number = machine.get("firmware_update_info", {}).get("id", None)
+            machine_type = machine.get("id", None)
+            if machine_bom_number and machine_type:
+                found_machine_type_identifiers[str(machine_bom_number)] = machine_type
+        return found_machine_type_identifiers
+
+    ## Add a new device.
+    def _onDeviceDiscovered(self, key: str, address: str, properties: Dict[bytes, bytes]) -> None:
+        cluster_size = int(properties.get(b"cluster_size", -1))
+        machine_identifier = properties.get(b"machine", b"").decode("utf-8")
+        printer_type_identifiers = self._getPrinterTypeIdentifiers()
+
+        # Detect the machine type based on the BOM number that is sent over the network.
+        properties[b"printer_type"] = b"Unknown"
+        for bom, p_type in printer_type_identifiers.items():
+            if machine_identifier.startswith(bom):
+                properties[b"printer_type"] = bytes(p_type, encoding="utf8")
+                break
+
+        # We no longer support legacy devices, so check that here.
+        if cluster_size == -1:
+            return
+
+        device = NetworkOutputDevice(key, address, properties)
+        CuraApplication.getInstance().getDiscoveredPrintersModel().addDiscoveredPrinter(
+            ip_address=address,
+            key=device.getId(),
+            name=device.getName(),
+            create_callback=self._createMachineFromDiscoveredDevice,
+            machine_type=device.printerType,
+            device=device
+        )
+        self._discovered_devices[device.getId()] = device
+        self.discoveredDevicesChanged.emit()
+        self._connectToActiveMachine()
+
+    ## Remove a device.
+    def _onDiscoveredDeviceRemoved(self, device_id: str) -> None:
+        device = self._discovered_devices.pop(device_id, None)  # type: Optional[NetworkOutputDevice]
+        if not device:
+            return
+        device.close()
+        CuraApplication.getInstance().getDiscoveredPrintersModel().removeDiscoveredPrinter(device.address)
+        self.discoveredDevicesChanged.emit()
+
+    ## Create a machine instance based on the discovered network printer.
```

Comment:
It's not an option to put this kind of logic in the Cura `OutputDeviceManager` right?

Reply:
Not really, because that global manager only wants devices in it that are for the active machine. These managers do specific stuff for discovery, local network, cloud networking, etc. The output device type is also specific. I also don't wanna touch core Cura code in this PR as that complicates the review.